File: tick/optim/prox/src/prox_l1w.cpp
#include "prox_l1w.h"

ProxL1w::ProxL1w(double strength,
                 SArrayDoublePtr weights,
                 bool positive)
    : ProxSeparable(strength, positive) {
    this->weights = weights;
}

ProxL1w::ProxL1w(double strength,
                 SArrayDoublePtr weights,
                 ulong start,
                 ulong end,
                 bool positive)
    : ProxSeparable(strength, start, end, positive) {
    this->weights = weights;
}

const std::string ProxL1w::get_class_name() const {
    return "ProxL1w";
}

void ProxL1w::call_single(ulong i,
                          const ArrayDouble &coeffs,
                          double step,
                          ArrayDouble &out) const {
    double thresh = step * strength;
    double coeffs_i = coeffs[i];
    double thresh_i = thresh * (*weights)[i];
    if (coeffs_i > 0) {
        if (coeffs_i > thresh_i) {
            out[i] = coeffs_i - thresh_i;
        } else {
            out[i] = 0;
        }
    } else {
        // If coeffs_i is negative we set it to 0
        if (positive) {
            out[i] = 0;
        } else {
            if (coeffs_i < -thresh_i) {
                out[i] = coeffs_i + thresh_i;
            } else {
                out[i] = 0;
            }
        }
    }
}

void ProxL1w::call_single(ulong i,
                          const ArrayDouble &coeffs,
                          double step,
                          ArrayDouble &out,
                          ulong n_times) const {
    if (n_times >= 1) {
        call_single(i, coeffs, n_times * step, out);
    } else {
        out[i] = coeffs[i];
    }
}

double ProxL1w::value_single(ulong i,
                             const ArrayDouble &coeffs) const {
    return (*weights)[i] * std::abs(coeffs[i]);
}

File: tick/optim/prox/prox_l1w.py
# License: BSD 3 clause

# -*- coding: utf8 -*-

import numpy as np
from .base import Prox
from .build.prox import ProxL1w as _ProxL1w

__author__ = 'Stephane Gaiffas'


# TODO: if we set a weights vector with length != end - start ???


class ProxL1w(Prox):
    """Proximal operator of the weighted L1 norm (weighted
    soft-thresholding)

    Parameters
    ----------
    strength : `float`
        Level of L1 penalization

    weights : `numpy.ndarray`, shape=(n_coeffs,)
        The weights to be used in the L1 penalization

    range : `tuple` of two `int`, default=`None`
        Range on which the prox is applied. If `None` then the prox is
        applied on the whole vector

    positive : `bool`, default=`False`
        If True, apply L1 penalization together with a projection
        onto the set of vectors with non-negative entries
    """

    _attrinfos = {
        "strength": {
            "writable": True,
            "cpp_setter": "set_strength"
        },
        "weights": {
            "writable": True,
            "cpp_setter": "set_weights"
        },
        "positive": {
            "writable": True,
            "cpp_setter": "set_positive"
        }
    }

    def __init__(self, strength: float, weights: np.ndarray,
                 range: tuple=None, positive: bool=False):
        Prox.__init__(self, range)
        if range is None:
            self._prox = _ProxL1w(strength, weights, positive)
        else:
            start, end = range
            if (end - start) != weights.shape[0]:
                raise ValueError("Size of ``weights`` does not match "
                                 "the given ``range``")
            self._prox = _ProxL1w(strength, weights, range[0], range[1],
                                  positive)
        self.positive = positive
        self.strength = strength
        self.weights = weights

    def _call(self, coeffs: np.ndarray, step: object, out: np.ndarray):
        self._prox.call(coeffs, step, out)

    def value(self, coeffs: np.ndarray):
        """
        Returns the value of the penalization at ``coeffs``

        Parameters
        ----------
        coeffs : `numpy.ndarray`, shape=(n_coeffs,)
            The value of the penalization is computed at this point

        Returns
        -------
        output : `float`
            Value of the penalization at ``coeffs``
        """
        return self._prox.value(coeffs)

    def _as_dict(self):
        dd = Prox._as_dict(self)
        del dd["weights"]
        return dd
